### ISSUE TITLE
clojure.test/is doesn't work inside then

### ISSUE THREAD

**Issue by artemyarulin** [funcool/promesa#25]
I've found out quiet strange behaviour of promesa and `is` macro when using Clojure version, consider those tests:

``` clojure
#?(:cljs
   (deftest cljc-p-test
     (async done
            (p/then (p/promise 42) (fn [_]
                                     (is (= 2 1))
                                     (done))))))
#?(:clj
   (deftest cljc-p-test
     @(p/then (p/promise 42) (fn [_]
                               (is (= 2 1))))))

```

ClojureScript version fails like it should, but Clojure test always passes. If I evaluate following line `@(p/then (p/promise 42) (fn [_] (is (= 2 1))))` it will return `false` without throwing anything, it will be resolved.

I'm not sure what may cause such thing, I'll appreciate any help.

**Comment by artemyarulin** [comment]
So `(macroexpand-all (quote (p/then (p/promise 42) (fn [_] (is (= 2 1)))))` returns this:

``` clojure
(clojure.core/deref
 (promesa.core/then
  (promesa.core/promise 42)
  (fn* ([_]
        (try
          (let* [values__7924__auto__ (clojure.core/list 2 1)
                 result__7925__auto__ (clojure.core/apply clojure.core/= values__7924__auto__)]
            (if result__7925__auto__
              (clojure.test/do-report {:type :pass,
                                       :expected (quote (clojure.core/= 2 1)),
                                       :actual (clojure.core/cons clojure.core/= values__7924__auto__),
                                       :message nil})
              (clojure.test/do-report {:type :fail,
                                       :expected (quote (clojure.core/= 2 1))
                                       :actual (clojure.core/list (quote not) (clojure.core/cons (quote clojure.core/=) values__7924__auto__))
                                       :message nil}))
            result__7925__auto__)
          (catch java.lang.Throwable t__7952__auto__
            (clojure.test/do-report {:type :error
                                     :expected (quote (clojure.core/= 2 1))
                                     :actual t__7952__auto__
                                     :message nil})))))))
```

If I would try to evaluate this `fn*` function manually in REPL and pass any dummy parameter then whole things throws like it should

**Comment by artemyarulin** [comment]
Did some test with `core.async` and closest example like this 

``` clojure
#?(:clj
   (deftest cljc-p-a-test
     (a/<!! (let [out (a/chan 1 (map (fn[_] (is (= 2 1)))))]
              (a/onto-chan out [1])))))
```

Fails like it should

**Comment by artemyarulin** [comment]
OK, looks like test reporting doesn't work from inside promesa block:

``` clojure
(defn r [_]
  (clojure.test/do-report {:type :fail,
                           :expected (quote false),
                           :actual false,
                           :message nil}))
(r 1) ;; logs to repl error message
(p/then (p/promise 1) r) ;; doesn't print anything
```

**Comment by artemyarulin** [comment]
Update, here simple Clojure code which never fails:

``` clojure

(clojure.test/deftest b
     (let [o (promise)]
       (.start (Thread. (fn []
                          (clojure.test/is false)
                          (deliver o 1))))
       @o))
```

Going though clojure.test, I've found out that code could be replaced with 

``` clojure
(clojure.test/deftest b
     (let [o (promise)]
       (.start (Thread. (fn []
                          (clojure.test/inc-report-counter :false)
                          (deliver o 1))))
       @o))
```
where `inc-report-counter` relying on dynamic variables a lot.

After some discussions on #clojure channels I was told (and I noticed myself) that core.async works as expected. Here the source from it related to that (I guess):

``` clojure
(defn thread-call
  "Executes f in another thread, returning immediately to the calling
  thread. Returns a channel which will receive the result of calling
  f when completed."
  [f]
  (let [c (chan 1)]
    (let [binds (clojure.lang.Var/getThreadBindingFrame)]
      (.execute thread-macro-executor
                (fn []
                  (clojure.lang.Var/resetThreadBindingFrame binds)
                  (try
                    (let [ret (f)]
                      (when-not (nil? ret)
                        (>!! c ret)))
                    (finally
                      (close! c))))))
    c))
```

Another useful quotes: 

```

joost-diepenmaat [11:11 AM]  
I think async/thread does some capturing/rebinding of the current dynamic environment

joost-diepenmaat [11:16 AM]  
before 1.3 all thread creation fns lost their dynamic bindings. AFAIK that doesn’t happen anymore unless you use the java Thread API directly.
```

**Comment by artemyarulin** [comment]
Unfortunately my JVM kung-fu is too weak to make a change by myself :(

I hope you would be able to help

**Comment by niwinz** [comment]
Thanks for your report.  I'm currently on my vacations, so I take it asap
in some days.

**Comment by artemyarulin** [comment]
Enjoy your vacation, it's not time critical for me at that time. And thank you again for the great library

**Comment by artemyarulin** [comment]
Hi, I'm about to start integration of Promesa library to my code base and I wonder what do you think - will it be fixed in near time?

**Comment by niwinz** [comment]
Hmm after some though about it,  it is expected behavior, because as you have noticed yourself, promesa is does not aware of the dynamic context thread locals that clojure have. On my tests I resolve all promises synchronously in order to test its return value and that was enough for me.

Although, seems like it is pretty easy to fix, just seeing the code snippets from core.async. It simple needs some additional code in promise chaining functions in order to capture the clojure dynamic context frame and set it inside the new thread.  In summary: it should be fixed pretty quickly and easily, but I considere this more like a feature that a bug.

**Comment by artemyarulin** [comment]
Cool, any time estimation when you would have time?

**Comment by niwinz** [comment]
`1.5.0-SNAPSHOT` published with that change, please check if it works as expected :D

**Comment by artemyarulin** [comment]
Oh yeah, I'll check tonight, thank you!

**Comment by artemyarulin** [comment]
Fixed! Looking forward to see a release in clojars :)

**Comment by niwinz** [comment]
Released 1.5.0 ;)

Thank you very much for report this!
